source/simulator/introspection.cc

```diff
@@ -32,166 +32,157 @@ namespace aspect
       else
-        fes.push_back(new FE_Q<dim>(parameters.composition_degree));
-      multiplicities.push_back(parameters.n_compositional_fields);
-      bes.compositional_fields = base_element++;
-
-      Assert(base_element == fes.size(), ExcInternalError());
-      Assert(base_element == multiplicities.size(), ExcInternalError());
-
-      return std_cxx11::make_tuple(bes, fes, multiplicities);
+        return std_cxx11::shared_ptr<FiniteElement<dim> >(new FE_Q<dim>(degree));
     }
-
-    /**
-     * Construct mapping from component to block indices.
-     */
     template <int dim>
-    std::vector<unsigned int>
-    setup_component_to_blocks (const typename Introspection<dim>::ComponentIndices &component_indices,
-                               const typename Introspection<dim>::BlockIndices &block_indices,
-                               const unsigned int n_components)
+    std_cxx11::shared_ptr<FiniteElement<dim> >
+    new_FE_Q_or_DGQ(const bool discontinuous,
+                    const unsigned int degree)
     {
-      std::vector<unsigned int> components_to_blocks;
-      const unsigned int n_compositional_fields = component_indices.compositional_fields.size();
-
-      components_to_blocks.resize(n_components, dealii::numbers::invalid_unsigned_int);
-      for (unsigned int d=0; d<dim; ++d)
-        components_to_blocks[component_indices.velocities[d]] = block_indices.velocities;
-      components_to_blocks[component_indices.pressure] = block_indices.pressure;
-      components_to_blocks[component_indices.temperature] = block_indices.temperature;
-      for (unsigned int c=0; c<n_compositional_fields; ++c)
-        components_to_blocks[component_indices.compositional_fields[c]] = block_indices.compositional_fields[c];
-
-#ifdef DEBUG
-      // check we assigned all components
-      for (unsigned int c=0; c<n_components; ++c)
-        Assert(components_to_blocks[c]!=dealii::numbers::invalid_unsigned_int, ExcInternalError());
-#endif
-
-      return components_to_blocks;
+      if (discontinuous)
+        return std_cxx11::shared_ptr<FiniteElement<dim> >(new FE_DGQ<dim>(degree));
+      else
+        return std_cxx11::shared_ptr<FiniteElement<dim> >(new FE_Q<dim>(degree));
     }
   }
 
 
 
   template <int dim>
-  Introspection<dim>::Introspection(const Parameters<dim> &parameters)
+  std::vector<dealii::VariableDefinition<dim> >
+  Introspection<dim>::construct_variables (const Parameters<dim> &parameters)
+  {
+    std::vector<dealii::VariableDefinition<dim> > variables;
+    bool use_direct_solver = parameters.use_direct_stokes_solver;
+
+
+    // u
+    variables.push_back(
+      dealii::VariableDefinition<dim>("velocity",
+                                      std_cxx11::shared_ptr<FiniteElement<dim> >(
+                                        new FE_Q<dim>(parameters.stokes_velocity_degree)),
+                                      dim,
+                                      use_direct_solver?0:1)
```

Comment:
We miscommunicated. I didn't object to the creation of a variable inside the call to `push_back` -- that's elegant, and should stay that way. My objection was to the piece `use_direct_solver?0:1` because that is coming out of blue air and at first seems to have nothing to do at all with the creation of a finite element. Might be better to write it as
```
  bool some_descriptive_name = use_direct_solver?0:1;
  variables.push_back (....., some_descriptive_name));
```